File: src/hooks/__tests__/agentStateMachine.test.ts
// Unit tests for agentStateMachine — state transitions
import { describe, it, expect } from "vitest";
import {
  createAgentState,
  applySignal,
  applyScreenStatus,
  notifyOutput,
  tick,
  setCliType,
  resetAgentState,
  type AgentState,
} from "../agentStateMachine";

describe("createAgentState", () => {
  it("starts in unknown status", () => {
    const state = createAgentState(1000);
    expect(state.status).toBe("unknown");
    expect(state.cli).toBe("unknown");
    expect(state.blockedMessage).toBeNull();
  });
});

describe("applySignal — blocked", () => {
  it("transitions to blocked on Devin notify signal", () => {
    const state = createAgentState(1000);
    applySignal(state, { kind: "blocked", cli: "devin", message: "Needs input" }, 2000);
    expect(state.status).toBe("blocked");
    expect(state.blockedMessage).toBe("Needs input");
    expect(state.cli).toBe("devin");
  });

  it("blocked is a priority state — fires immediately", () => {
    const state = createAgentState(1000);
    // Even from unknown, blocked fires immediately
    applySignal(state, { kind: "blocked", cli: "devin", message: "Approval" }, 2000);
    expect(state.status).toBe("blocked");
  });
});

describe("applySignal — done", () => {
  it("transitions to done on devin-idle signal", () => {
    const state = createAgentState(1000);
    // First go to working
    notifyOutput(state, 2000);
    state.pendingStatus = null;
    state.status = "working"; // force past debounce for test
    // Then done
    applySignal(state, { kind: "done", cli: "devin" }, 3000);
    expect(state.status).toBe("done");
    expect(state.blockedMessage).toBeNull();
  });
});

describe("applySignal — notify (OSC 777/9)", () => {
  it("transitions to done on 'Devin finished' notify signal", () => {
    const state = createAgentState(1000);
    state.cli = "devin";
    state.status = "working";
    applySignal(state, { kind: "notify", cli: "devin", message: "Devin finished", done: true }, 2000);
    expect(state.status).toBe("done");
    expect(state.blockedMessage).toBeNull();
  });

  it("transitions to blocked on 'Devin needs input' notify signal", () => {
    const state = createAgentState(1000);
    state.cli = "devin";
    state.status = "working";
    applySignal(state, { kind: "notify", cli: "devin", message: "Devin needs input", done: false }, 2000);
    expect(state.status).toBe("blocked");
    expect(state.blockedMessage).toBe("Devin needs input");
  });

  it("transitions to done on 'Devin encountered an error' notify signal", () => {
    const state = createAgentState(1000);
    state.cli = "devin";
    state.status = "working";
    applySignal(state, { kind: "notify", cli: "devin", message: "Devin encountered an error", done: true }, 2000);
    expect(state.status).toBe("done");
  });

  it("blocked → done on 'Devin finished' notify (user answered + Devin finished)", () => {
    const state = createAgentState(1000);
    state.cli = "devin";
    applySignal(state, { kind: "notify", cli: "devin", message: "Devin needs input", done: false }, 2000);
    expect(state.status).toBe("blocked");
    applySignal(state, { kind: "notify", cli: "devin", message: "Devin finished", done: true }, 3000);
    expect(state.status).toBe("done");
  });
});

describe("notifyOutput", () => {
  it("does NOT transition unknown → working when CLI is unknown (plain shell)", () => {
    const state = createAgentState(1000);
    notifyOutput(state, 2000);
    // CLI is unknown — plain shell output should NOT trigger working
    expect(state.status).toBe("unknown");
    expect(state.pendingStatus).toBeNull();
  });

  it("transitions to working (debounced) when CLI is detected and status is unknown", () => {
    const state = createAgentState(1000);
    state.cli = "devin";
    notifyOutput(state, 2000);
    // Debounced — status hasn't changed yet
    expect(state.status).toBe("unknown");
    expect(state.pendingStatus).toBe("working");
  });

  it("transitions idle → working (debounced) when CLI is detected", () => {
    const state = createAgentState(1000);
    state.cli = "devin";
    state.status = "idle";
    notifyOutput(state, 2000);
    // Debounced — status hasn't changed yet
    expect(state.status).toBe("idle");
    expect(state.pendingStatus).toBe("working");
  });

  it("fires debounced working after tick", () => {
    const state = createAgentState(1000);
    state.cli = "devin";
    state.status = "idle";
    notifyOutput(state, 2000);
    expect(state.pendingStatus).toBe("working");
    // Tick after debounce window (500ms)
    tick(state, 2600);
    expect(state.status).toBe("working");
    expect(state.pendingStatus).toBeNull();
  });

  it("does NOT transition blocked → working on output (spinner animation)", () => {
    const state = createAgentState(1000);
    applySignal(state, { kind: "blocked", cli: "devin", message: "Approval" }, 2000);
    expect(state.status).toBe("blocked");
    // PTY output during blocked (spinner, cursor blink) should NOT clear blocked
    notifyOutput(state, 3000);
    expect(state.status).toBe("blocked");
    expect(state.blockedMessage).toBe("Approval");
  });

  it("updates lastOutputAt on every output", () => {
    const state = createAgentState(1000);
    notifyOutput(state, 2000);
    expect(state.lastOutputAt).toBe(2000);
    notifyOutput(state, 2500);
    expect(state.lastOutputAt).toBe(2500);
  });
});

describe("tick — done → idle decay", () => {
  it("decays done → idle after 5s of no output", () => {
    const state = createAgentState(1000);
    // Get to done state
    state.status = "done";
    state.lastOutputAt = 1000;
    // Tick at 4001ms — not yet 5s
    tick(state, 4001);
    expect(state.status).toBe("done");
    // Tick at 6001ms — past 5s
    tick(state, 6001);
    expect(state.status).toBe("idle");
  });

  it("does not decay if output happened recently", () => {
    const state = createAgentState(1000);
    state.status = "done";
    state.lastOutputAt = 5000;
    tick(state, 8000); // only 3s since last output
    expect(state.status).toBe("done");
  });
});

describe("tick — working → done timeout (fallback)", () => {
  it("transitions working → done after 60s of no output (CLI detected)", () => {
    const state = createAgentState(1000);
    state.cli = "devin";
    state.status = "working";
    state.lastOutputAt = 1000;
    // Tick at 61001ms — past 60s timeout (61001 - 1000 = 60001 > 60000)
    tick(state, 61001);
    expect(state.status).toBe("done");
  });

  it("does NOT transition working → done if output happened recently", () => {
    const state = createAgentState(1000);
    state.cli = "devin";
    state.status = "working";
    state.lastOutputAt = 50000;
    tick(state, 60000); // only 10s since last output
    expect(state.status).toBe("working");
  });

  it("does NOT transition working → done if CLI is unknown (plain shell)", () => {
    const state = createAgentState(1000);
    state.cli = "unknown";
    state.status = "working";
    state.lastOutputAt = 1000;
    tick(state, 70000); // 69s since last output, but no CLI
    expect(state.status).toBe("working");
  });

  it("does NOT transition blocked → done (blocked is sticky)", () => {
    const state = createAgentState(1000);
    state.cli = "devin";
    state.status = "blocked";
    state.lastOutputAt = 1000;
    tick(state, 70000);
    expect(state.status).toBe("blocked");
  });
});

describe("tick — debounce firing", () => {
  it("fires pending working after debounce window", () => {
    const state = createAgentState(1000);
    state.cli = "devin";
    state.status = "idle";
    notifyOutput(state, 2000);
    // pendingStatus = working, pendingFireAt = 2500
    tick(state, 2400);
    expect(state.status).toBe("idle"); // not yet
    tick(state, 2600);
    expect(state.status).toBe("working");
  });
});

describe("setCliType", () => {
  it("sets CLI type without changing status", () => {
    const state = createAgentState(1000);
    state.status = "working";
    setCliType(state, "claude-code");
    expect(state.cli).toBe("claude-code");
    expect(state.status).toBe("working");
  });
});

describe("applyScreenStatus", () => {
  it("blocked sets status immediately + stores message", () => {
    const state = createAgentState(1000);
    state.status = "working";
    applyScreenStatus(state, "blocked", "Permission required", 2000);
    expect(state.status).toBe("blocked");
    expect(state.blockedMessage).toBe("Permission required");
    expect(state.pendingStatus).toBeNull();
  });

  it("blocked clears blockedMessage when message is null", () => {
    const state = createAgentState(1000);
    state.blockedMessage = "old message";
    applyScreenStatus(state, "blocked", null, 2000);
    expect(state.status).toBe("blocked");
    expect(state.blockedMessage).toBeNull();
  });

  it("working transitions immediately (no debounce)", () => {
    const state = createAgentState(1000);
    state.status = "idle";
    applyScreenStatus(state, "working", null, 2000);
    // working from screen scrape is immediate — no pendingStatus
    expect(state.status).toBe("working");
    expect(state.pendingStatus).toBeNull();
  });

  it("done transitions immediately + clears blockedMessage", () => {
    const state = createAgentState(1000);
    state.status = "blocked";
    state.blockedMessage = "need input";
    applyScreenStatus(state, "done", null, 2000);
    expect(state.status).toBe("done");
    expect(state.blockedMessage).toBeNull();
    expect(state.pendingStatus).toBeNull();
  });

  it("idle does NOT override blocked", () => {
    const state = createAgentState(1000);
    state.status = "blocked";
    state.blockedMessage = "Permission required";
    applyScreenStatus(state, "idle", null, 2000);
    expect(state.status).toBe("blocked");
    expect(state.blockedMessage).toBe("Permission required");
  });

  it("idle transitions from non-blocked status", () => {
    const state = createAgentState(1000);
    state.status = "working";
    applyScreenStatus(state, "idle", null, 2000);
    expect(state.status).toBe("idle");
    expect(state.pendingStatus).toBeNull();
  });

  it("idle does NOT override done (screen scrape must not skip the 5s done→idle decay)", () => {
    // When Devin finishes, OSC sets status="done" (blue dot). The screen
    // immediately shows the idle placeholder (e.g. "❭ Ask Devin to...").
    // Screen scrape detects "idle" but must NOT override "done" — otherwise
    // the blue dot flashes for <1 tick and instantly turns gray, and the
    // user never sees that Devin finished. The done→idle decay is handled
    // by the tick timer (DONE_TO_IDLE_MS = 5000ms), not by screen scrape.
    const state = createAgentState(1000);
    state.status = "done";
    state.lastOutputAt = 1000;
    applyScreenStatus(state, "idle", null, 2000);
    expect(state.status).toBe("done");
    // After 5s tick decay, done→idle fires
    tick(state, 6001);
    expect(state.status).toBe("idle");
  });

  it("idle DOES transition from working (no done state to protect)", () => {
    // When there's no OSC done signal (CLI doesn't emit OSC), screen scrape
    // detecting the idle placeholder should transition working→idle directly.
    const state = createAgentState(1000);
    state.status = "working";
    applyScreenStatus(state, "idle", null, 2000);
    expect(state.status).toBe("idle");
  });

  it("idle DOES transition from unknown (initial detection)", () => {
    // When CLI is first detected from screen and the screen shows the idle
    // placeholder, status should transition from unknown→idle.
    const state = createAgentState(1000);
    state.status = "unknown";
    applyScreenStatus(state, "idle", null, 2000);
    expect(state.status).toBe("idle");
  });
});

describe("resetAgentState", () => {
  it("resets to initial state", () => {
    const state = createAgentState(1000);
    state.status = "blocked";
    state.cli = "devin";
    state.blockedMessage = "test";
    state.blockedFromOsc = true;
    resetAgentState(state, 5000);
    expect(state.status).toBe("unknown");
    expect(state.cli).toBe("unknown");
    expect(state.blockedMessage).toBeNull();
    expect(state.blockedFromOsc).toBe(false);
  });
});

describe("State transition edge cases", () => {
  it("done → working on new output (new turn)", () => {
    const state = createAgentState(1000);
    state.cli = "devin";
    state.status = "done";
    notifyOutput(state, 2000);
    expect(state.pendingStatus).toBe("working");
    tick(state, 2600);
    expect(state.status).toBe("working");
  });

  it("idle → working on new output", () => {
    const state = createAgentState(1000);
    state.cli = "devin";
    state.status = "idle";
    notifyOutput(state, 2000);
    expect(state.pendingStatus).toBe("working");
    tick(state, 2600);
    expect(state.status).toBe("working");
  });

  it("blocked → done on devin-idle signal", () => {
    const state = createAgentState(1000);
    applySignal(state, { kind: "blocked", cli: "devin", message: "Approval" }, 2000);
    expect(state.status).toBe("blocked");
    applySignal(state, { kind: "done", cli: "devin" }, 3000);
    expect(state.status).toBe("done");
  });
});

File: android/app/src/test/java/com/termfast/app/agent/AgentStateMachineTest.kt
package com.termfast.app.agent

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue

class AgentStateMachineTest {

    @Test
    fun testCreateAgentState() {
        val state = AgentStateMachine.createAgentState(1000L)
        assertEquals(AgentStatus.UNKNOWN, state.status)
        assertEquals(CliType.UNKNOWN, state.cli)
        assertEquals(1000L, state.lastOutputAt)
        assertEquals(1000L, state.lastStatusChangeAt)
        assertNull(state.pendingStatus)
    }

    // === applySignal ===

    @Test
    fun testApplySignalBlocked() {
        val state = AgentStateMachine.createAgentState(1000L)
        state.cli = CliType.DEVIN
        val signal = AgentSignal.Blocked(CliType.DEVIN, "needs input")
        AgentStateMachine.applySignal(state, signal, 2000L)
        assertEquals(AgentStatus.BLOCKED, state.status)
        assertEquals("needs input", state.blockedMessage)
        assertTrue(state.blockedFromOsc)
    }

    @Test
    fun testApplySignalDone() {
        val state = AgentStateMachine.createAgentState(1000L)
        state.cli = CliType.DEVIN
        state.status = AgentStatus.WORKING
        val signal = AgentSignal.Done(CliType.DEVIN)
        AgentStateMachine.applySignal(state, signal, 2000L)
        assertEquals(AgentStatus.DONE, state.status)
    }

    @Test
    fun testApplySignalNotifyDone() {
        val state = AgentStateMachine.createAgentState(1000L)
        state.cli = CliType.DEVIN
        state.status = AgentStatus.WORKING
        val signal = AgentSignal.Notify(CliType.DEVIN, "Devin finished", true)
        AgentStateMachine.applySignal(state, signal, 2000L)
        assertEquals(AgentStatus.DONE, state.status)
    }

    @Test
    fun testApplySignalNotifyBlocked() {
        val state = AgentStateMachine.createAgentState(1000L)
        state.cli = CliType.DEVIN
        state.status = AgentStatus.WORKING
        val signal = AgentSignal.Notify(CliType.DEVIN, "Devin needs input", false)
        AgentStateMachine.applySignal(state, signal, 2000L)
        assertEquals(AgentStatus.BLOCKED, state.status)
    }

    @Test
    fun testApplySignalTitleUpdatesCli() {
        val state = AgentStateMachine.createAgentState(1000L)
        val signal = AgentSignal.Title(CliType.OPENCODE, "OpenCode")
        AgentStateMachine.applySignal(state, signal, 2000L)
        assertEquals(CliType.OPENCODE, state.cli)
    }

    @Test
    fun testApplySignalNullReturnsCurrentStatus() {
        val state = AgentStateMachine.createAgentState(1000L)
        state.status = AgentStatus.WORKING
        val status = AgentStateMachine.applySignal(state, null, 2000L)
        assertEquals(AgentStatus.WORKING, status)
    }

    // === applyScreenStatus ===

    @Test
    fun testApplyScreenStatusBlocked() {
        val state = AgentStateMachine.createAgentState(1000L)
        AgentStateMachine.applyScreenStatus(state, AgentStatus.BLOCKED, "question?", 2000L)
        assertEquals(AgentStatus.BLOCKED, state.status)
        assertEquals("question?", state.blockedMessage)
        assertFalse(state.blockedFromOsc)
    }

    @Test
    fun testApplyScreenStatusWorking() {
        val state = AgentStateMachine.createAgentState(1000L)
        state.status = AgentStatus.IDLE
        AgentStateMachine.applyScreenStatus(state, AgentStatus.WORKING, null, 2000L)
        assertEquals(AgentStatus.WORKING, state.status)
    }

    @Test
    fun testApplyScreenStatusIdleDoesNotOverrideBlocked() {
        val state = AgentStateMachine.createAgentState(1000L)
        state.status = AgentStatus.BLOCKED
        AgentStateMachine.applyScreenStatus(state, AgentStatus.IDLE, null, 2000L)
        assertEquals(AgentStatus.BLOCKED, state.status)
    }

    @Test
    fun testApplyScreenStatusDone() {
        val state = AgentStateMachine.createAgentState(1000L)
        state.status = AgentStatus.WORKING
        AgentStateMachine.applyScreenStatus(state, AgentStatus.DONE, null, 2000L)
        assertEquals(AgentStatus.DONE, state.status)
        assertNull(state.blockedMessage)
    }

    // === clearScreenBlocked ===

    @Test
    fun testClearScreenBlocked() {
        val state = AgentStateMachine.createAgentState(1000L)
        state.status = AgentStatus.BLOCKED
        state.blockedMessage = "msg"
        state.blockedMissCount = 3
        AgentStateMachine.clearScreenBlocked(state, AgentStatus.WORKING, 2000L)
        assertEquals(AgentStatus.WORKING, state.status)
        assertNull(state.blockedMessage)
        assertEquals(0, state.blockedMissCount)
    }

    @Test
    fun testClearScreenBlockedNoOpIfNotBlocked() {
        val state = AgentStateMachine.createAgentState(1000L)
        state.status = AgentStatus.WORKING
        AgentStateMachine.clearScreenBlocked(state, AgentStatus.IDLE, 2000L)
        assertEquals(AgentStatus.WORKING, state.status)
    }

    // === notifyOutput ===

    @Test
    fun testNotifyOutputWhenIdleSchedulesWorking() {
        val state = AgentStateMachine.createAgentState(1000L)
        state.cli = CliType.DEVIN
        state.status = AgentStatus.IDLE
        AgentStateMachine.notifyOutput(state, 2000L)
        assertEquals(AgentStatus.WORKING, state.pendingStatus)
    }

    @Test
    fun testNotifyOutputWhenBlockedStaysBlocked() {
        val state = AgentStateMachine.createAgentState(1000L)
        state.cli = CliType.DEVIN
        state.status = AgentStatus.BLOCKED
        AgentStateMachine.notifyOutput(state, 2000L)
        assertEquals(AgentStatus.BLOCKED, state.status)
        assertEquals(2000L, state.lastOutputAt)
    }

    @Test
    fun testNotifyOutputWhenCliUnknownNoChange() {
        val state = AgentStateMachine.createAgentState(1000L)
        state.status = AgentStatus.IDLE
        AgentStateMachine.notifyOutput(state, 2000L)
        assertEquals(AgentStatus.IDLE, state.status)
    }

    // === tick (debounce + done→idle decay) ===

    @Test
    fun testTickFiresPendingTransition() {
        val state = AgentStateMachine.createAgentState(1000L)
        state.cli = CliType.DEVIN
        state.status = AgentStatus.IDLE
        // Schedule a debounced transition to WORKING
        AgentStateMachine.notifyOutput(state, 1000L)
        assertEquals(AgentStatus.WORKING, state.pendingStatus)
        // Tick before debounce window → no transition yet
        AgentStateMachine.tick(state, 1200L)
        assertEquals(AgentStatus.IDLE, state.status)
        // Tick after debounce window (500ms) → transition fires
        AgentStateMachine.tick(state, 1600L)
        assertEquals(AgentStatus.WORKING, state.status)
        assertNull(state.pendingStatus)
    }

    @Test
    fun testTickDoneToIdleDecay() {
        val state = AgentStateMachine.createAgentState(1000L)
        state.status = AgentStatus.DONE
        state.lastOutputAt = 1000L
        // Tick before decay timeout (5s) → stays DONE
        AgentStateMachine.tick(state, 4000L)
        assertEquals(AgentStatus.DONE, state.status)
        // Tick after decay timeout → transitions to IDLE
        AgentStateMachine.tick(state, 7000L)
        assertEquals(AgentStatus.IDLE, state.status)
    }

    // === setCliType ===

    @Test
    fun testSetCliType() {
        val state = AgentStateMachine.createAgentState(1000L)
        AgentStateMachine.setCliType(state, CliType.OPENCODE)
        assertEquals(CliType.OPENCODE, state.cli)
    }

    // === resetAgentState ===

    @Test
    fun testResetAgentState() {
        val state = AgentStateMachine.createAgentState(1000L)
        state.status = AgentStatus.WORKING
        state.cli = CliType.DEVIN
        state.blockedMessage = "msg"
        AgentStateMachine.resetAgentState(state, 5000L)
        assertEquals(AgentStatus.UNKNOWN, state.status)
        assertEquals(CliType.UNKNOWN, state.cli)
        assertNull(state.blockedMessage)
    }
}
